docs/architecture/adr-022-multi-versioned-state-machine.md

```diff
@@ -0,0 +1,98 @@
+# ADR 022: Multi-Versioned State Machine
+
+## Changelog
+
+- 2024/03/15: Initial draft (@cmwaters)
+
+## Status
+
+Implemented
+
+## Context
+
+The Celestia application required a modification from the existing Cosmos SDK to support multiple versions of the state machine simultaneously. This capability is crucial for single binary syncs and ensuring a smooth transitions during upgrades, allowing nodes to upgrade independently and then switch to the next state machine without any downtime to the network. This is important for a network which is depended on by a large number of rollups and users.
+
+## Decision
+
+We decided to implement a multi-versioned state machine by making significant modifications to the module manager, the configurator, and introducing a new ante handler. These changes enable the application to maintain and execute different versions of the state machine based on the app version associated with each transaction and prevents the execution of transactions that are not part of the current state machine.
```

Comment:
> based on the app version associated with each transaction

is the app version associated with a transaction or with a block? I think the latter.

Reply:
There is a mapping between transaction type and what versions it can be executed during. If a block is a version and includes a transaction not belonging to that version it is considered invalid

Reply:
Oh I see what this was saying. On re-read looks fine, can resolve.